### PR TITLE
add UniversalDocumentProcessor's input model

### PR DESCRIPTION
**Pull request by superyyrrzz**

### CODE REVIEW

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/InheritanceTree.cs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.UniversalReference
+{
+    using System;
+    using System.Collections.Generic;
+
+    [Serializable]
+    public class InheritanceTree : List<InheritanceTree>
+    {
```
Comment (vwxyzh) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/InheritanceTree.cs]: no member?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/UniversalReferenceDocumentProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+                {
+                    yield return GetXRefInfo(reference, key);
+                }
+            }
+        }
+
+        private static XRefSpec GetXRefInfo(ReferenceViewModel item, string key)
+        {
+            var result = GetXRefSpecFromReference(item);
+            result.Href = ((RelativePath)key).UrlEncode().ToString();
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/UniversalReferenceDocumentProcessor.cs]: >Href [](start = 19, length = 4)

question: is there any purpose not encode href in GetXRefSpecFromReference?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/MemberType.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Build.UniversalReference
+{
+    using System;
+
+    [Serializable]
+    public enum MemberType
```
Comment (vwxyzh) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/MemberType.cs]: maybe not enough for FP language or dynamic language.

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+        [JsonProperty(Constants.PropertyName.CommentId)]
+        public string CommentId { get; set; }
+
+        [YamlMember(Alias = Constants.PropertyName.Id)]
+        [JsonProperty(Constants.PropertyName.Id)]
+        public string Id { get; set; }
+
+        [YamlMember(Alias = "isEii")]
+        [JsonProperty("isEii")]
+        public bool IsExplicitInterfaceImplementation { get; set; }
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs]: >IsExplicitInterfaceImplementation [](start = 20, length = 33)

is eii c# or vb specific?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/UniversalReferenceDocumentProcessor.cs (hunk 2 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/UniversalReferenceDocumentProcessor.cs]: item.Href is already encoded

---
In reply to: [112601786](https://github.com/dotnet/docfx/pull/1574#discussion_r112601786) [](ancestors = 112601786)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs (hunk 4 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs]: Yes, it should also apply for c# and vb

---
In reply to: [112602605](https://github.com/dotnet/docfx/pull/1574#discussion_r112602605) [](ancestors = 112602605)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/InheritanceTree.cs (hunk 1 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/InheritanceTree.cs]: oh... should add it

---
In reply to: [112601655](https://github.com/dotnet/docfx/pull/1574#discussion_r112601655) [](ancestors = 112601655)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+        [YamlMember(Alias = "isExtensionMethod")]
+        [JsonProperty("isExtensionMethod")]
+        public bool IsExtensionMethod { get; set; }
+
+        [YamlMember(Alias = Constants.PropertyName.Parent)]
+        [JsonProperty(Constants.PropertyName.Parent)]
+        [UniqueIdentityReference]
+        public string Parent { get; set; }
```
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs]: >string [](start = 15, length = 6)

List<strign>

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs (hunk 4 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs]: Agree it's too specific and can also be handled in metadata. remove

---
In reply to: [112603489](https://github.com/dotnet/docfx/pull/1574#discussion_r112603489) [](ancestors = 112603489,112602605)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/MemberType.cs (hunk 3 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/MemberType.cs]: change to string

---
In reply to: [112601880](https://github.com/dotnet/docfx/pull/1574#discussion_r112601880) [](ancestors = 112601880)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/InheritanceTree.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        /// <summary>
+        /// item's inheritance
+        /// multiple inheritance is allowed in languages like Python
+        /// </summary>
+        [YamlMember(Alias = Constants.PropertyName.Inheritance)]
+        [MergeOption(MergeOption.Ignore)]
+        [JsonProperty(Constants.PropertyName.Inheritance)]
+        public List<InheritanceTree> Inheritance { get; set; }
+
+    }
```
Comment (hellosnow) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/InheritanceTree.cs]: add some extensible metadata?

**Comment by hellosnow** [comment]
:shipit:

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/UniversalReferenceConstants.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        public const string UniversalReference = "UniversalReference";
+        public const string UniversalReferenceYamlMime = YamlMime.YamlMimePrefix + UniversalReference;
+
+        public static class PropertyName
+        {
+            public const string Returns = "returns";
+        }
+
+        public static class ExtensionMemberPrefix
```
Comment (vicancy) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/UniversalReferenceConstants.cs]: >ExtensionMemberPrefix [](start = 28, length = 21)

Move to Constants?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ApiParameter.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        /// </summary>
+        [YamlMember(Alias = "type")]
+        [JsonProperty("type")]
+        [UniqueIdentityReference]
+        public List<string> Type { get; set; }
+
+        [YamlMember(Alias = "description")]
+        [JsonProperty("description")]
+        [MarkdownContent]
+        public string Description { get; set; }
```
Comment (vicancy) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ApiParameter.cs]: >string [](start = 15, length = 6)

multiple types <=> multiple descriptions?

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/UniversalReferenceDocumentProcessor.cs [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+                case DocumentType.Article:
+                    if (".yml".Equals(Path.GetExtension(file.File), StringComparison.OrdinalIgnoreCase) ||
+                        ".yaml".Equals(Path.GetExtension(file.File), StringComparison.OrdinalIgnoreCase))
+                    {
+                        var mime = YamlMime.ReadMime(file.File);
+                        switch (mime)
+                        {
+                            case UniversalReferenceConstants.UniversalReferenceYamlMime:
+                                return ProcessingPriority.Normal;
+                            case null:
```
Comment (vwxyzh) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/UniversalReferenceDocumentProcessor.cs]: remove.

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ApiParameter.cs (hunk 8 above)
Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ApiParameter.cs]: it's parameter's description

---
In reply to: [112862172](https://github.com/dotnet/docfx/pull/1574#discussion_r112862172) [](ancestors = 112862172)

File: plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs (hunk 5 above)
Comment (vwxyzh) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs]: ?

---
In reply to: [112604924](https://github.com/dotnet/docfx/pull/1574#discussion_r112604924) [](ancestors = 112604924)

Comment (superyyrrzz) [inline plugins/Microsoft.DocAsCode.Build.UniversalReference/BuildInputs/ItemViewModel.cs]: should be string...

---
In reply to: [112870641](https://github.com/dotnet/docfx/pull/1574#discussion_r112870641) [](ancestors = 112870641,112604924)

**Pull request merged**